_posts/2025-09-05-monitoring-of-custom-resources.md

```diff
@@ -0,0 +1,536 @@
+---
+layout: post
+title:  "Monitoring of custom resources"
+date: 2025-09-05
+author: sebastian_gaiser_hetzner
+---
+
+Running Kafka on Kubernetes with Strimzi is straightforward.
+But getting a complete, consistent view of everything in your Kubernetes and Kafka cluster(s) can still be hard.
+Kafka topics and users are often owned by different teams and configured differently, especially when a Kafka cluster is shared across a company (for example, by multiple teams).
+As usage grows, gathering information and troubleshooting gets harder.
+
+This post shows a practical, low‑friction way to close monitoring gaps using [kube-state-metrics](https://github.com/kubernetes/kube-state-metrics) so you can track every Strimzi custom resource (CR) with a simple, consistent, Kubernetes‑native approach.
+
+### What Strimzi exposes today
+
+All Strimzi operators (Cluster, User, and Topic Operator) expose metrics in Prometheus-compatible format.
+However, only the Cluster Operator exposes a metric (`strimzi_resource_state`) that reports the state of custom resources (ready or not).
+
+It provides this metric for:
+
+- `Kafka`
+- `KafkaConnect`
+- `KafkaBridge`
+- `KafkaMirrorMaker2`
+- `KafkaRebalance`
+
+It does not provide this metric for:
+
+- `KafkaNodePool`
+- `KafkaConnector`
+- `StrimziPodSet`
+
+The User and Topic Operators also do not provide this metric for `KafkaUser` and `KafkaTopic`.
+
+During routine work, it's easy to introduce misconfigurations, from a simple typo in a topic setting to using a property that was removed in the last release.
+These small issues can trigger long debugging sessions. Often the fix is simple, but finding it is not.
+
+### The simple fix: use kube-state-metrics for CRDs
+
+Instead of relying on the limited `strimzi_resource_state` metric, you can use kube-state-metrics to cover _all_ Strimzi custom resources.
+
+kube-state-metrics can read any Kubernetes object, including custom resources, and export Info-style metrics with labels you control. This approach gives you:
+
+- A complete inventory of all Strimzi CRs
+- Consistent metric labels across teams
+- A Kubernetes‑native approach that fits existing tooling (Prometheus)
+
+In short, you increase monitoring coverage without changing Strimzi or writing custom exporters.
+
+### How to deploy
+
+In general, Strimzi provides plain Kubernetes manifests in the examples directory.
+
+#### Option A — Apply manifests
+
+In [Strimzi 0.48.0](https://github.com/strimzi/strimzi-kafka-operator/releases/tag/0.48.0), examples are included to deploy kube-state-metrics with the corresponding:
+
+- RBAC resources (`ServiceAccount`, `RoleBinding`, `ClusterRoleBinding`)
+- A `ConfigMap` for CR configuration
+- The `Deployment`
+- A `ServiceMonitor` for [Prometheus Operator](https://prometheus-operator.dev/) to scrape itself
+
+[Examples directory for kube-state-metrics](https://github.com/strimzi/strimzi-kafka-operator/tree/0.48.0/examples/metrics/kube-state-metrics).
```

Comment:
I'm assuming the kube-state-metrics examples will be including in all Strimzi versions going forwards. If that is the case I would suggest the following tweaks to wording and links to make that clear:
```suggestion
From [Strimzi 0.48.0](https://github.com/strimzi/strimzi-kafka-operator/releases/tag/0.48.0) onwards, examples are included to deploy kube-state-metrics with the corresponding:

- RBAC resources (`ServiceAccount`, `RoleBinding`, `ClusterRoleBinding`)
- A `ConfigMap` for CR configuration
- The `Deployment`
- A `ServiceMonitor` for [Prometheus Operator](https://prometheus-operator.dev/) to scrape itself

[Examples directory for kube-state-metrics](https://github.com/strimzi/strimzi-kafka-operator/tree/main/examples/metrics/kube-state-metrics).
```